File: backend/api/api_key.py

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from ..services.api_key_manager import (
    issue_api_key_for_user, revoke_api_key, list_user_api_keys, get_api_key_details, validate_api_key, update_api_key_last_used, reactivate_api_key, count_api_keys
)
from ..schemas.api_key import APIKeyCreate, APIKeyRead
from ..database import get_db
from typing import List

router = APIRouter()

@router.post("/issue/{user_id}", response_model=APIKeyRead)
def issue_key(user_id: str, db: Session = Depends(get_db)):
    api_key = issue_api_key_for_user(db, user_id)
    return api_key

@router.post("/revoke/{key}")
def revoke_key(key: str, db: Session = Depends(get_db)):
    return {"success": revoke_api_key(db, key)}

@router.get("/user/{user_id}", response_model=List[APIKeyRead])
def list_keys(user_id: str, db: Session = Depends(get_db)):
    return list_user_api_keys(db, user_id)

@router.get("/details/{key}", response_model=APIKeyRead)
def key_details(key: str, db: Session = Depends(get_db)):
    api_key = get_api_key_details(db, key)
    if not api_key:
        raise HTTPException(status_code=404, detail="API key not found")
    return api_key

@router.get("/validate/{key}")
def validate_key(key: str, user_id: str = None, db: Session = Depends(get_db)):
    return {"valid": validate_api_key(db, key, user_id)}

@router.post("/reactivate/{key}")
def reactivate_key(key: str, db: Session = Depends(get_db)):
    return {"success": reactivate_api_key(db, key)}

@router.get("/count")
def count_keys(user_id: str = None, active_only: bool = False, db: Session = Depends(get_db)):
    return {"count": count_api_keys(db, user_id, active_only)}

File: frontend/src/api/api_key.ts
// API utility for api_key endpoints matching backend/api/api_key.py

export async function issueApiKey(userId: string) {
  const response = await fetch(`/api/api_key/issue/${userId}`, {
    method: 'POST',
  });
  if (!response.ok) throw new Error('Failed to issue API key');
  return response.json();
}

export async function revokeApiKey(key: string) {
  const response = await fetch(`/api/api_key/revoke/${key}`, {
    method: 'POST',
  });
  if (!response.ok) throw new Error('Failed to revoke API key');
  return response.json();
}

export async function listUserApiKeys(userId: string) {
  const response = await fetch(`/api/api_key/user/${userId}`);
  if (!response.ok) throw new Error('Failed to list user API keys');
  return response.json();
}

export async function getApiKeyDetails(key: string) {
  const response = await fetch(`/api/api_key/details/${key}`);
  if (!response.ok) throw new Error('Failed to get API key details');
  return response.json();
}

export async function validateApiKey(key: string, userId?: string) {
  const url = userId ? `/api/api_key/validate/${key}?user_id=${userId}` : `/api/api_key/validate/${key}`;
  const response = await fetch(url);
  if (!response.ok) throw new Error('Failed to validate API key');
  return response.json();
}

export async function reactivateApiKey(key: string) {
  const response = await fetch(`/api/api_key/reactivate/${key}`, {
    method: 'POST',
  });
  if (!response.ok) throw new Error('Failed to reactivate API key');
  return response.json();
}

export async function countApiKeys(userId?: string, activeOnly: boolean = false) {
  let url = `/api/api_key/count?active_only=${activeOnly}`;
  if (userId) url += `&user_id=${userId}`;
  const response = await fetch(url);
  if (!response.ok) throw new Error('Failed to count API keys');
  return response.json();
}
